Implement: multiple save and load for mlflow registry

yleilawang opened the pull request:
1. Implemented save_multiple and load_multiple for mlflow registry
2. Test cases for implementation.

s0nicboOm commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -187,6 +189,39 @@ def load(
+        loaded_model = self.load(skeys=skeys, dkeys=dkeys, artifact_type="pyfunc")
+        if loaded_model is not None:
+            metadata = loaded_model.artifact.unwrap_python_model().metadata
+            dict_artifacts = loaded_model.artifact.unwrap_python_model().dict_artifacts
+            artifacts_dict = {}
+            for artifact in dict_artifacts.values():
+                artifact_data = ArtifactData(
+                    artifact=artifact.artifact, metadata=metadata, extras=None
+                )
+                dynamic_key = ":".join(artifact.dkeys)
+                artifacts_dict[dynamic_key] = artifact_data
+        else:
+            artifacts_dict = None
```
Can we just return the ArtifactKey instead?

s0nicboOm commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -238,6 +276,37 @@ def save(
+            mlflow ModelVersion instance
+        """
+        multiple_artifacts = CompositeModels(skeys=skeys, dict_artifacts=dict_artifacts, **metadata)
+        dkeys_list = multiple_artifacts.get_dkeys_list()
```
make this function private?

s0nicboOm commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -338,3 +407,63 @@ def __load_artifacts(
+        metadata (META_VT): Additional metadata associated with the artifacts.
+    """
+
+    def __init__(self, skeys: KEYS, dict_artifacts: dict[str, KeyedArtifact], **metadata: META_VT):
```
add __slots__ for class

s0nicboOm commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -187,6 +189,39 @@ def load(
+        loaded_model = self.load(skeys=skeys, dkeys=dkeys, artifact_type="pyfunc")
+        if loaded_model is not None:
+            metadata = loaded_model.artifact.unwrap_python_model().metadata
+            dict_artifacts = loaded_model.artifact.unwrap_python_model().dict_artifacts
```
can we add excption handling for this unwrapping

s0nicboOm reviewed (changes requested):
Also, update the version for pyproject.toml

ab93 commented:
> Also, update the version for pyproject.toml

@s0nicboOm we have not released any previous versions right?

yleilawang commented:
Added changes for:
1. Changed return type of load_multiple to ArtifactData, the same return type of load.
2. Slots for CompositeModels class.
3. Handling exceptions caused by unwarpping pyfunc models.
4. Private function for getting unique sorted dkey list.
5. Added test cases for cache loading for pyfunc models.

yleilawang commented:
**Profiling**

`def f():
    reg.load(skeys=skeys, dkeys= ["ae", "pipeline"], artifact_type='pytorch')
    reg.load(skeys=skeys, dkeys= ["scaler", "pipeline"], artifact_type='sklearn')
    reg.load(skeys=skeys, dkeys= ["threshold", "pipeline"], artifact_type='sklearn')

def g():
    reg.load_multiple(skeys=skeys, dkeys_list=[["ae", "pipeline"], ["scaler", "pipeline"], ["threshold", "pipeline"]])

import timeit
a = timeit.timeit(f, number=10)
b = timeit.timeit(g, number=10)

print(a, b)`

1.0356064579991653 0.6088340830010566

ab93 commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -338,3 +406,65 @@ def __load_artifacts(
+        -------
+            list[list[str]]: A list of all dynamic keys in the stored artifacts.
+        """
+        dkeys_list = []
+        artifacts = self.dict_artifacts.values()
+        for artifact in artifacts:
+            dkeys_list.append(artifact.dkeys)
+        return dkeys_list
```
```suggestion
        return [ka.dkeys for _, ka  in self.dict_artifacts]
```

yleilawang replied:
Method not needed.

ab93 commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -338,3 +406,65 @@ def __load_artifacts(
+        return list(SortedSet([dkey for dkeys in dkeys_list for dkey in dkeys]))
+
+
+class CompositeModels(mlflow.pyfunc.PythonModel):
```
```suggestion
class CompositeModel(mlflow.pyfunc.PythonModel):
```

yleilawang commented:
**Profiling Updates**

```
reg.save(skeys=skeys, dkeys= ["ae", "pipeline"], artifact=VanillaAE(10), artifact_type='pytorch', **{"a": "b"})
reg.save(skeys=skeys, dkeys= ["scaler", "pipeline"], artifact=StandardScaler(), artifact_type='sklearn', **{"a": "b"})
reg.save(skeys=skeys, dkeys= ["threshold", "pipeline"], artifact=StdDevThreshold(), artifact_type='sklearn', **{"a": "b"})

output = reg.save_multiple(
    skeys=skeys, dkeys=dkeys, dict_artifacts=dict_artifacts, **{"a": "b"}
)

def f():
    reg.load(skeys=skeys, dkeys= ["ae", "pipeline"], artifact_type='pytorch')
    reg.load(skeys=skeys, dkeys= ["scaler", "pipeline"], artifact_type='sklearn')
    reg.load(skeys=skeys, dkeys= ["threshold", "pipeline"], artifact_type='sklearn')

def g():
    reg.load_multiple(skeys=skeys, dkeys=dkeys)

import timeit

a = timeit.timeit(f, number=100)
b = timeit.timeit(g, number=100)

print(a, b)
```
--> 9.074686333002319 4.829918750001525

ab93 commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -187,6 +188,43 @@ def load(
+        except Exception:
+            _LOGGER.exception("Unexpected error occurred while unwrapping python model.")
+            return None
+        else:
+            return ArtifactData(
+                artifact=unwrapped_composite_model.dict_artifacts,
+                metadata=loaded_model.metadata,
+                extras=loaded_model.extras,
+            )
```
```suggestion
        return ArtifactData(
            artifact=unwrapped_composite_model.dict_artifacts,
            metadata=loaded_model.metadata,
             extras=loaded_model.extras,
        )
```

ab93 commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -238,6 +279,40 @@ def save(
+
+        """
+        if len(dict_artifacts) == 1:
+            _LOGGER.warning("Only one element in dict_artifacts. Please use save directly.")
```
```suggestion
            _LOGGER.warning("Only one element in dict_artifacts. Saving directly is recommended.")
```

ab93 commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -338,3 +413,45 @@ def __load_artifacts(
+        self.dict_artifacts = dict_artifacts
+        self.metadata = metadata
+
+    def predict(self):
```
Can you match the signature of the base class here? I think it takes in some parameters

ab93 reviewed (approved):
Looks good!

s0nicboOm commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -18,6 +18,7 @@
 import mlflow.pyfunc
 import mlflow.pytorch
 import mlflow.sklearn
```
I think we don't need these imports?

yleilawang replied:
Test cases still need these imports. Otherwise will cause connection errors in tests.

ab93 replied:
Agreed. Let's optimize the imports

ab93 replied:
@yleilawang a good rule of thumb is to make the test cases to work around the actual code. Not the other way round! :)

yleilawang replied:
Can we move these imports to test file? Just tried and it didn't cause errors. @ab93 @s0nicboOm

ab93 replied:
Yes test file sounds good!

s0nicboOm commented on numalogic/registry/mlflow_registry.py:
```diff
@@ -238,6 +279,40 @@ def save(
+
+        """
+        if len(dict_artifacts) == 1:
+            _LOGGER.warning("Only one element in dict_artifacts. Saving directly is recommended.")
```
`_LOGGER.warning("Only one artifact present in dict_artifacts. Saving directly is recommended.")`

s0nicboOm reviewed (approved):
Minor comments. 
Otherwise LGTM!
